File: 42. Trapping Rain Water.py
# Problem Link: https://leetcode.com/problems/trapping-rain-water/


# Solution Using Dynamic Programming

class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) < 3:
            return 0
        left_tracker = [0] * len(height)
        right_tracker = [0] * len(height)
        m = 0
        for i in range(1, len(height)):
            m = max(m, height[i - 1])
            left_tracker[i] = m
        m = 0
        for i in range(len(height) - 2, 0, -1):
            m = max(m, height[i + 1])
            right_tracker[i] = m
        ans = 0
        for i in range(len(height)):
            if min(left_tracker[i], right_tracker[i]) > height[i]:
                ans += min(left_tracker[i], right_tracker[i]) - height[i]
        return ans

File: Leetcode/42. Trapping Rain Water.java
/* Problem link: https://leetcode.com/problems/trapping-rain-water/ */

// Approach 1: Dynamic Programming Time Complexity O(n), Space Complexity O(n)

class Solution {
    public int trap(int[] height) {
        int area = 0;
        int[] left = height.clone();
        int[] right = height.clone();
        for (int i = 1; i < height.length; i++) left[i] = Math.max(height[i], left[i - 1]);
        for (int i = height.length - 2; i >= 0; i--) right[i] = Math.max(height[i], right[i + 1]);
        for (int i = 1; i < height.length - 1; i++) {
            int min = Math.min(left[i], right[i]);
            if (height[i] < min) {
                area += min - height[i];
            }
        }
        return area;
    }
}

// Approach 2: Two Pointers: Time Complexity o(n), Space Complexity O(1)

class Solution {
    public int trap(int[] height) {
        int left_max = 0, right_max = 0;
        int left = 0, right = height.length - 1;
        int ans = 0;
        while (left < right) {
            left_max = Math.max(left_max, height[left]);
            right_max = Math.max(right_max, height[right]);
            if (left_max < right_max) {
                ans += Math.max(0, left_max - height[left]);
                left++;
            } else {
                ans += Math.max(0, right_max - height[right]);
                right--;
            }
        }
        return ans;
    }
}
